infer/component.go

```diff
@@ -30,40 +30,72 @@ import (
 	"github.com/pulumi/pulumi-go-provider/middleware/schema"
 )
 
+// ComponentResource may be turned into an [InferredComponent] with [Component].
+type ComponentResource[I any, O pulumi.ComponentResource] interface {
+	// Construct a component resource
+	//
+	// ctx.RegisterResource needs to be called, but ctx.RegisterOutputs does not need to
+	// be called.
+	Construct(ctx *pulumi.Context, name, typ string, args I, opts pulumi.ResourceOption) (O, error)
+}
+
 // InferredComponent is a component resource inferred from code.
 //
-// To create an [InferredComponent], call the [Component] function.
+// To create an [InferredComponent], call the [ComponentF] function.
 type InferredComponent interface {
 	t.ComponentResource
 	schema.Resource
 
 	isInferredComponent()
 }
 
-func (derivedComponentController[R, I, O]) isInferredComponent() {}
+func (derivedComponentController[R, T, I, O]) isInferredComponent() {}
+
+// Component defines a component resource from go code. Here `R` is the component resource
+// anchor, `I` describes its inputs and `O` its outputs. To add descriptions to `R`, `I`
+// and `O`, see the `Annotated` trait defined in this module.
+func Component[R ComponentResource[I, O], I any, O pulumi.ComponentResource](rsc R) InferredComponent {
+	return &derivedComponentController[R, R, I, O]{receiver: &rsc}
+}
 
 // ComponentFn describes the type signature of a Pulumi component resource defined in Go.
 type ComponentFn[A any, R pulumi.ComponentResource] = func(
 	ctx *pulumi.Context, name string, args A, opts ...pulumi.ResourceOption,
 ) (R, error)
 
-// Component creates an [InferredComponent] using functions and types that a existing Pulumi component program
+type componentF[I any, O pulumi.ComponentResource] struct {
+	construct ComponentFn[I, O]
+}
+
+func (fn *componentF[I, O]) Construct(ctx *pulumi.Context, name, typ string, inputs I, opts pulumi.ResourceOption,
+) (O, error) {
+	return fn.construct(ctx, name, inputs, opts)
+}
+
+// ComponentF creates an [InferredComponent] using functions and types that an existing Pulumi component program
 // would have implemented.
 //
 // fn is the function you would use to construct the program.
 //
+// Note that this syntax doesn't allow annotating the component. Use
+// [infer.Component] for more control over the component's generated schema.
```

Comment:
Before this PR, it did allow annotating the component.

Reply:
I agree this is untrue because annotations may be applied to the component resource; I'll remove this comment.